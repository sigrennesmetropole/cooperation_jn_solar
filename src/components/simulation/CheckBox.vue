<script setup lang="ts">
import { computed, ref, watchEffect } from 'vue'
import type { Ref } from 'vue'

const props = defineProps<{
  isOnError: boolean
  isChecked: boolean
}>()

const emit = defineEmits<{
  (e: 'checkBoxChange', status: boolean): void
}>()

const checked = ref(props.isChecked)
const isFocus: Ref<boolean> = ref(false)

watchEffect(() => {
  checked.value = props.isChecked
})

function statusChange() {
  checked.value = !checked.value

  emit('checkBoxChange', checked.value)
}

const borderCheckbox = computed(() => {
  if (isFocus.value) {
    return 'border-2 border-black'
  }
  if (props.isOnError && !checked.value) {
    return 'border-red-600'
  }
  if (checked.value) {
    return 'border-slate-900'
  }
  return 'border-slate-400'
})
</script>

<template>
  <div class="flex p-0 gap-3 items-top">
    <div
      class="shadow-sm w-5 h-5 rounded border flex justify-center items-center mt-[3px]"
      :class="borderCheckbox"
      @click="statusChange()"
      @keyup.enter="statusChange()"
    >
      <input
        type="checkbox"
        class="opacity-0 absolute"
        @focus="isFocus = true"
        @blur="isFocus = false"
      />
      <svg
        class="w-3 h-3 pointer-events-none stroke-slate-900 stroke-2"
        viewBox="0 0 14 14"
        :style="checked ? '' : 'display: none;'"
      >
        <path
          d="M0.5 8.5498L3.23 12.0598C3.32212 12.1795 3.44016 12.2768 3.57525 12.3443C3.71034 12.4119 3.85898 12.4479 4.01 12.4498C4.15859 12.4515 4.3057 12.4201 4.44063 12.3578C4.57555 12.2956 4.6949 12.204 4.79 12.0898L13.5 1.5498"
          fill="none"
        />
      </svg>
    </div>
    <div class="flex w-fit">
      <slot name="text"></slot>
    </div>
  </div>
</template>
