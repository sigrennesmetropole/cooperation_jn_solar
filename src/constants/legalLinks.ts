import type { LegalItem } from '@sigrennesmetropole/cooperation_jn_common_ui/dist/model/legalItem.model'

export const legalList: LegalItem[] = [
  {
    link: '/legalnotice/mentions-legales',
    slug: 'mentions-legales',
    name: 'Mentions légales',
  },
  {
    link: '/legalnotice/confidentialite',
    slug: 'confidentialite',
    name: 'Politique de confidentialité',
  },
  {
    link: '/legalnotice/plan-du-site',
    slug: 'plan-du-site',
    name: 'Plan du site',
  },
  {
    link: '/legalnotice/accessibilite',
    slug: 'accessibilite',
    name: 'Accessibilité : non conforme',
  },
  {
    link: '/legalnotice/cookies',
    slug: 'cookies',
    name: 'Cookies',
  },
]
