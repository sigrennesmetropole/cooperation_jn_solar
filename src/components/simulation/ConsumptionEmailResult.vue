<script setup lang="ts">
import CheckBox from '@/components/simulation/CheckBox.vue'
import expandIcon from '../../assets/icons/expand-small-bigger-retract-smaller-big.svg'
import deleteCircle from '../../assets/icons/interface-delete-circle.svg'
import { computed, ref } from 'vue'
import { legalList } from '@/constants/legalLinks'
// @ts-ignore : Could not find a declaration file for module 'dompurify'
import DOMPurify from 'dompurify'

const validEmail = ref(true)
const isCheckBoxOnError = ref(false)

const userEmail = ref('')
const isCheckboxChecked = ref(false)

const errorPicto = computed(() => {
  if (validEmail.value) {
    return 'hidden'
  } else {
    return 'visible absolute right-3 top-10'
  }
})

function checkboxChange(event: boolean) {
  isCheckboxChecked.value = event
  changeError()
}

function changeError() {
  isCheckBoxOnError.value = !isCheckboxChecked.value
}

function sendEmail() {
  const email = DOMPurify.sanitize(userEmail.value)
  const emailRegex =
    /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
  validEmail.value = emailRegex.test(email)

  changeError()

  if (validEmail.value && !isCheckBoxOnError.value) {
    console.log('send consumption by email')
  }
}

const openPrivacy = () => {
  window.open(
    legalList.find((elt) => elt.slug === 'confidentialite')!.link,
    '_blank'
  )
}
</script>

<template>
  <div
    class="flex flex-col gap-8 px-6 py-12 w-[576px] h-fit bg-slate-50 rounded-xl border border-slate-200"
  >
    <div class="flex flex-col gap-3 text-center">
      <h4 class="text-xl font-bold">Vous ne souhaitez pas attendre ?</h4>
      <p>
        Renseignez votre mail pour recevoir<br />les résultats dès qu'ils seront
        prêts :
      </p>
    </div>
    <div class="flex flex-col gap-6">
      <div class="flex flex-col gap-1.5 relative">
        <label class="font-normal">Votre adresse mail</label>
        <div :class="errorPicto">
          <img :src="deleteCircle" alt="" />
        </div>
        <input
          type="email"
          v-model="userEmail"
          class="border border-neutral-400 rounded"
        />
        <div
          id="errorEmail"
          v-if="!validEmail"
          class="bg-rose-50 border border-rose-200 rounded-lg p-2 text-rose-900 text-sm font-normal"
        >
          Veuillez vérifier votre adresse mail.
        </div>
      </div>

      <CheckBox
        :isOnError="isCheckBoxOnError"
        :isChecked="false"
        @checkBoxChange="checkboxChange($event)"
      >
        <template v-slot:text>
          <p class="font-normal text-base">
            J'autorise la transmission de mon adresse mail pour l'envoi de ma
            simulation. Celle-ci ne sera pas conservée.
          </p>
        </template>
      </CheckBox>

      <div class="flex flex-row text-neutral-700">
        <p>
          Vous pouvez lire notre
          <span
            tabindex="0"
            @click="openPrivacy()"
            @keydown.enter="openPrivacy()"
            class="font-medium underline decoration-1 hover:cursor-pointer"
            data-testid="privacy-link"
          >
            Politique de confidentialité
          </span>
        </p>
        <img :src="expandIcon" class="w-3.5 h-3.5 ml-2 mt-1" />
      </div>
      <div class="flex flex-row justify-center">
        <button
          class="bg-black shadow-sm rounded-lg gap-3 px-4 py-3 items-center flex flex-row justify-center w-fit"
          @click="sendEmail()"
        >
          <span class="text-white text-base font-medium"
            >Recevoir mon rapport en PDF</span
          >
        </button>
      </div>
    </div>
  </div>
</template>
