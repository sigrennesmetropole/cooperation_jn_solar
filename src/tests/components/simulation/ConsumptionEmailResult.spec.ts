import { mount } from '@vue/test-utils'
import ConsumptionEmailResult from '@/components/simulation/ConsumptionEmailResult.vue'
import CheckBox from '@/components/simulation/CheckBox.vue'
import { legalList } from '@/constants/legalLinks'

describe('ConsumptionEmailResult', () => {
  it('renders correctly', () => {
    const wrapper = mount(ConsumptionEmailResult)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('validates email input correctly', async () => {
    const wrapper = mount(ConsumptionEmailResult)
    const emailInput = wrapper.find('input[type="email"]')

    await emailInput.setValue('invalid-email')
    await wrapper.find('button').trigger('click')
    await wrapper.vm.$nextTick()

    expect(wrapper.vm.validEmail).toBe(false)
    expect(wrapper.find('#errorEmail').exists()).toBe(true)
    expect(wrapper.find('#errorEmail').classes()).toContain('bg-rose-50')
    expect(wrapper.find('#errorEmail').text()).toBe(
      'Veuillez vÃ©rifier votre adresse mail.'
    )

    await emailInput.setValue('valid-email@example.com')
    await wrapper.find('button').trigger('click')
    await wrapper.vm.$nextTick()

    expect(wrapper.vm.validEmail).toBe(true)
    expect(wrapper.find('#errorEmail').exists()).toBe(false)
  })

  it('updates isCheckboxChecked when CheckBox emits checkBoxChange', async () => {
    const wrapper = mount(ConsumptionEmailResult)
    const checkBox = wrapper.findComponent(CheckBox)

    await checkBox.vm.$emit('checkBoxChange', true)
    expect(wrapper.vm.isCheckboxChecked).toBe(true)

    await checkBox.vm.$emit('checkBoxChange', false)
    expect(wrapper.vm.isCheckboxChecked).toBe(false)
  })

  it('calls openPrivacy when privacy link is clicked', async () => {
    const wrapper = mount(ConsumptionEmailResult)
    const privacyLink = wrapper.find('[data-testid="privacy-link"]')

    // @ts-ignore
    const windowOpenSpy = jest
      .spyOn(window, 'open')
      .mockImplementation(() => {})

    await privacyLink.trigger('click')
    const privacyLinkData = legalList.find(
      (elt) => elt.slug === 'confidentialite'
    )
    // @ts-ignore
    expect(windowOpenSpy).toHaveBeenCalledWith(privacyLinkData.link, '_blank')

    windowOpenSpy.mockRestore()
  })
})
