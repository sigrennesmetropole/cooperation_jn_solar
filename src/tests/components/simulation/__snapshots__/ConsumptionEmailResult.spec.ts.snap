// Vitest Snapshot v1

exports[`ConsumptionEmailResult > renders correctly 1`] = `
"<div class=\\"flex flex-col gap-8 px-6 py-12 w-[576px] h-fit bg-slate-50 rounded-xl border border-slate-200\\">
  <div class=\\"flex flex-col gap-3 text-center\\">
    <h4 class=\\"text-xl font-bold\\">Vous ne souhaitez pas attendre ?</h4>
    <p> Renseignez votre mail pour recevoir<br>les résultats dès qu'ils seront prêts : </p>
  </div>
  <div class=\\"flex flex-col gap-6\\">
    <div class=\\"flex flex-col gap-1.5 relative\\"><label class=\\"font-normal\\">Votre adresse mail</label>
      <div class=\\"hidden\\"><img src=\\"/src/assets/icons/interface-delete-circle.svg\\" alt=\\"\\"></div><input type=\\"email\\" class=\\"border border-neutral-400 rounded\\">
      <!--v-if-->
    </div>
    <div class=\\"flex p-0 gap-3 items-top\\">
      <div class=\\"shadow-sm w-5 h-5 rounded border flex justify-center items-center mt-[3px] border-slate-400\\"><input type=\\"checkbox\\" class=\\"opacity-0 absolute\\"><svg class=\\"w-3 h-3 pointer-events-none stroke-slate-900 stroke-2\\" viewBox=\\"0 0 14 14\\" style=\\"display: none;\\">
          <path d=\\"M0.5 8.5498L3.23 12.0598C3.32212 12.1795 3.44016 12.2768 3.57525 12.3443C3.71034 12.4119 3.85898 12.4479 4.01 12.4498C4.15859 12.4515 4.3057 12.4201 4.44063 12.3578C4.57555 12.2956 4.6949 12.204 4.79 12.0898L13.5 1.5498\\" fill=\\"none\\"></path>
        </svg></div>
      <div class=\\"flex w-fit\\">
        <p class=\\"font-normal text-base\\"> J'autorise la transmission de mon adresse mail pour l'envoi de ma simulation. Celle-ci ne sera pas conservée. </p>
      </div>
    </div>
    <div class=\\"flex flex-row text-neutral-700\\">
      <p> Vous pouvez lire notre <span tabindex=\\"0\\" class=\\"font-medium underline decoration-1 hover:cursor-pointer\\" data-testid=\\"privacy-link\\"> Politique de confidentialité </span></p><img src=\\"/src/assets/icons/expand-small-bigger-retract-smaller-big.svg\\" class=\\"w-3.5 h-3.5 ml-2 mt-1\\">
    </div>
    <div class=\\"flex flex-row justify-center\\"><button class=\\"bg-black shadow-sm rounded-lg gap-3 px-4 py-3 items-center flex flex-row justify-center w-fit\\"><span class=\\"text-white text-base font-medium\\">Recevoir mon rapport en PDF</span>
        <!--v-if-->
      </button></div>
    <!--v-if-->
  </div>
</div>"
`;
